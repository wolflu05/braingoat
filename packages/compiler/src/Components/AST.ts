import { Braingoat, ErrorType } from "../Braingoat";
import { findIndexAt, isValidVariableName } from "../utils";
import { LineType, TokenType } from "./Tokenizer";

export enum TOKEN_TYPES {
  DECLARATION = "DECLARATION",
  FUNCTION_CALL = "FUNCTION_CALL",
  EXPRESSION = "EXPRESSION",
  VALUE_LITERAL = "VALUE_LITERAL",
  ASSIGNMENT = "ASSIGNMENT",
  VARIABLE_LITERAL = "VARIABLE_LITERAL",
}

export type DECLARATION_OPTIONS = {
  variableName: string;
  type: string;
  value: AST;
};

export type FUNCTION_CALL_OPTIONS = {
  functionName: string;
  parameters: TokenType[];
  block: AST[];
};

export enum EXPRESSION_OPERATOR_TYPE {
  ADD = "ADD",
  SUBTRACT = "SUBTRACT",
  MULTIPLY = "MULTIPLY",
  DIVIDE = "DIVIDE",
  POWER = "POWER",
  EQ = "EQ",
  NEQ = "NEQ",
}
export const OPERATOR_MAP = {
  "+": EXPRESSION_OPERATOR_TYPE.ADD,
  "-": EXPRESSION_OPERATOR_TYPE.SUBTRACT,
  "*": EXPRESSION_OPERATOR_TYPE.MULTIPLY,
  "/": EXPRESSION_OPERATOR_TYPE.DIVIDE,
  "^": EXPRESSION_OPERATOR_TYPE.POWER,
  "==": EXPRESSION_OPERATOR_TYPE.EQ,
  "!=": EXPRESSION_OPERATOR_TYPE.NEQ,
};
export type EXPRESSION_OPTIONS = {
  op: EXPRESSION_OPERATOR_TYPE;
  leftNode: AST;
  rightNode: AST;
};

export type VALUE_LITERAL_OPTIONS =
  | {
      type: "INT";
      value: number;
    }
  | {
      type: "INTLIST";
      value: number[];
    };
export type VARIABLE_LITERAL_OPTIONS = {
  name: string;
};

export type ASSIGNMENT_OPTIONS = {
  variableName: string;
  value: AST;
};

export type TOKEN_OPTIONS =
  | DECLARATION_OPTIONS
  | FUNCTION_CALL_OPTIONS
  | EXPRESSION_OPTIONS
  | VALUE_LITERAL_OPTIONS
  | VARIABLE_LITERAL_OPTIONS
  | ASSIGNMENT_OPTIONS;

export class AST {
  type: TOKEN_TYPES;
  tokenOptions: TOKEN_OPTIONS;
  source: LineType;

  constructor(type: TOKEN_TYPES, tokenOptions: TOKEN_OPTIONS, source: LineType) {
    this.type = type;
    this.tokenOptions = tokenOptions;
    this.source = source;
  }

  static parseExpression(tokens: TokenType[], braingoat: Braingoat, startIndex = 0): [AST, number] | null {
    const token = tokens[startIndex];
    if (!token) {
      return null;
    }

    if (token.value === "(") {
      const exprIndex = [];
      const stack = [];
      let k = startIndex;

      for (; k < tokens.length; k++) {
        if (stack.length === 1) {
          exprIndex.push(k);
        }

        if (tokens[k].value === "(") {
          stack.push("(");
        } else if (tokens[k].value === ")") {
          stack.pop();
        }

        if (stack.length === 0) break;
      }

      if (stack.length !== 0) {
        braingoat.throwError(ErrorType.SyntaxError, "Expected )", tokens[k - 1]);
      }

      if (exprIndex.length !== 4) {
        braingoat.throwError(
          ErrorType.SyntaxError,
          "Invalid length expression, expressions can only have two operants wrapped in brackets",
          tokens[k - 1],
        );
      }

      const op = OPERATOR_MAP[tokens[exprIndex[1]].value as keyof typeof OPERATOR_MAP];
      const leftNode = this.parseExpression(tokens, braingoat, exprIndex[0]);
      const rightNode = this.parseExpression(tokens, braingoat, exprIndex[2]);

      if (!op) {
        braingoat.throwError(
          ErrorType.SyntaxError,
          `Invalid operator ${tokens[exprIndex[1]].value}`,
          tokens[exprIndex[1]],
        );
      }
      if (!leftNode) {
        braingoat.throwError(ErrorType.SyntaxError, "Expected left-side operant", tokens[exprIndex[1] - 1]);
      }
      if (!rightNode) {
        braingoat.throwError(ErrorType.SyntaxError, "Expected right-side operant", tokens[exprIndex[1]]);
      }

      return [
        new AST(
          TOKEN_TYPES.EXPRESSION,
          {
            op,
            leftNode: leftNode?.[0] || null,
            rightNode: rightNode?.[0] || null,
          },
          tokens[exprIndex[1]].source,
        ),
        k + 1,
      ];
    }

    if (!Number.isNaN(+token.value)) {
      return [new AST(TOKEN_TYPES.VALUE_LITERAL, { type: "INT", value: +token.value }, token.source), startIndex + 1];
    }

    if (isValidVariableName(token.value)) {
      return [new AST(TOKEN_TYPES.VARIABLE_LITERAL, { name: token.value }, token.source), startIndex + 1];
    }

    return null;
  }

  static parse(tokens: TokenType[], braingoat: Braingoat) {
    const tree: AST[] = [];

    for (let i = 0; i < tokens.length; ) {
      let nextIndex = i;

      // DECLARATION
      if (tokens[i + 2]?.value === "=") {
        const expression = this.parseExpression(tokens, braingoat, i + 3);
        if (!expression) {
          braingoat.throwError(ErrorType.SyntaxError, "Invalid expression", tokens[i + 3]);
        }

        tree.push(
          new AST(
            TOKEN_TYPES.DECLARATION,
            {
              variableName: tokens[i + 1].value,
              type: tokens[i].value,
              value: expression[0],
            },
            tokens[i].source,
          ),
        );
        nextIndex = expression[1];
      }

      // ASSIGNMENT
      else if (tokens[i + 1]?.value === "=") {
        const expression = this.parseExpression(tokens, braingoat, i + 2);
        if (!expression) {
          braingoat.throwError(ErrorType.SyntaxError, "Invalid expression", tokens[i + 2]);
        }

        tree.push(
          new AST(
            TOKEN_TYPES.ASSIGNMENT,
            {
              variableName: tokens[i].value,
              value: expression[0],
            },
            tokens[i].source,
          ),
        );
        nextIndex = expression[1];
      }

      // FUNCTION_CALL
      else if (tokens[i + 1]?.value === "(") {
        const endTokenIndex = findIndexAt(i + 1, tokens, (t) => t.value === ")");
        if (endTokenIndex === -1) {
          braingoat.throwError(ErrorType.SyntaxError, "Expected )", tokens[i + 1]);
        }

        const params = tokens.slice(i + 2, endTokenIndex);
        const wrongIdx = params.findIndex((v, i) => ((i + 1) % 2 === 0 ? v.value !== "," : false));

        if (wrongIdx !== -1) {
          braingoat.throwError(ErrorType.SyntaxError, `Expected ,`, params[wrongIdx]);
        }

        // check for block function
        let block: TokenType[] = [];
        if (tokens[endTokenIndex + 1]?.value === "{" && tokens.length > endTokenIndex + 2) {
          const stack = [];
          let k = endTokenIndex + 1;

          for (; k < tokens.length; k++) {
            if (tokens[k].value === "{") {
              stack.push("{");
            } else if (tokens[k].value === "}") {
              stack.pop();
            }

            if (stack.length === 0) break;
            block.push(tokens[k]);
          }
          if (stack.length !== 0) {
            braingoat.throwError(ErrorType.SyntaxError, "Expected }", tokens[k - 1]);
          }

          nextIndex = k + 1;
        } else {
          nextIndex = endTokenIndex + 1;
        }

        tree.push(
          new AST(
            TOKEN_TYPES.FUNCTION_CALL,
            {
              functionName: tokens[i].value,
              parameters: params.filter((x) => x.value !== ","),
              block: AST.parse(block.slice(1), braingoat),
            },
            tokens[i].source,
          ),
        );
      }

      // UNPROCESSED KEYWORD
      else {
        braingoat.throwError(ErrorType.SyntaxError, "Unprocessed keyword", tokens[i]);
      }

      if (i === nextIndex) {
        nextIndex++;
      }
      i = nextIndex;
    }

    return tree;
  }
}
