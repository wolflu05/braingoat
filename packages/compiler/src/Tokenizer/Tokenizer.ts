import { generateSplitRegex } from "../utils";
import { OPERATOR_MAP } from "../AST";

export type LineType = {
  line: number;
  col: number;
};

export type TokenType = {
  value: string;
  source: LineType;
};

export class Tokenizer {
  static tokenize(code: string[]) {
    const tokens: TokenType[] = [];
    const splitters = [" ", "(", ")", "{", "}", "[", "]", ",", "<", ">", "/*", "*/", ...Object.keys(OPERATOR_MAP), "="];

    for (const [lineNum, line] of Object.entries(code)) {
      const re = generateSplitRegex(splitters);
      let m, p;
      while (((p = re.lastIndex), (m = re.exec(line)))) {
        // This is necessary to avoid infinite loops with zero-width matches
        if (m.index === re.lastIndex) {
          re.lastIndex++;
        }

        // token between match
        tokens.push({
          value: line.substring(p, m.index),
          source: { line: +lineNum, col: p },
        });

        // actual matched token
        tokens.push({
          value: m[0],
          source: { line: +lineNum, col: m.index },
        });
      }

      // last token in line
      tokens.push({
        value: line.substring(p),
        source: { line: +lineNum, col: p },
      });
    }

    return tokens.filter((token) => !["", " ", "\n"].includes(token.value));
  }
}
