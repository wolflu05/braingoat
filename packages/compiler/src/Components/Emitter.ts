import { Braingoat, ErrorType } from "../Braingoat";
import { DataTypes } from "../DataTypes";
import { Int } from "../DataTypes/Int";
import { Functions } from "../Functions";
import { BasicDataType, memoryPositionType } from "../Types/BasicDataType";
import { COMMANDS, VALID_CHARS } from "../utils/commands";
import { AST, DECLARATION_OPTIONS, FUNCTION_CALL_OPTIONS, TOKEN_TYPES } from "./AST";
import { LineType, TokenType } from "./Tokenizer";

export class Emitter {
  memoryAllocation: BasicDataType[];
  currentMemoryAddress: number;
  code: string;
  ast: AST[];
  braingoat: Braingoat;

  constructor(ast: AST[], braingoat: Braingoat) {
    this.memoryAllocation = [];
    this.currentMemoryAddress = 0;
    this.code = "";
    this.ast = ast;
    this.braingoat = braingoat;
  }

  emit() {
    for (const node of this.ast) {
      // VARIABLE DECLARATION
      if (node.type === TOKEN_TYPES.DECLARATION) {
        const { type, variableName, value } = node.tokenOptions as DECLARATION_OPTIONS;
        this.registerVariable(type, variableName, value, node.source);
      }

      // FUNCTION CALL
      else if (node.type === TOKEN_TYPES.FUNCTION_CALL) {
        const { functionName, parameters, block } = node.tokenOptions as FUNCTION_CALL_OPTIONS;
        this.callFunction(functionName, parameters, block, node.source);
      }
    }

    return this;
  }

  movePointerTo(address: number, addToCode = true) {
    let res = "";

    while (this.currentMemoryAddress !== address) {
      if (this.currentMemoryAddress < address) {
        res += COMMANDS.MOVE_POINTER_RIGHT;
        this.currentMemoryAddress++;
      } else {
        res += COMMANDS.MOVE_POINTER_LEFT;
        this.currentMemoryAddress--;
      }
    }

    if (addToCode) this.addCode(res);

    return res;
  }

  getNextNEmpty(length: number): memoryPositionType {
    let pos = this.memoryAllocation
      .sort((a, b) => a.position[0] - b.position[0])
      .findIndex((_, i, l) => {
        if (!l[i + 1]) return true;
        return l[i + 1].position[0] - l[i].position[1] - 1 >= length;
      });

    if (pos === -1) {
      pos = 0;
    } else {
      pos = this.memoryAllocation[pos].position[1] + 1;
    }

    return [pos, pos + length - 1];
  }

  registerVariable(type: string, name: string, value: string, source: LineType) {
    if (!(type in DataTypes)) {
      this.braingoat.throwError(ErrorType.CompileError, `${type} is no valid data type`, source);
    }

    this.memoryAllocation.push(new DataTypes[type as keyof typeof DataTypes](this, name, value, source));
  }

  getVariable(name: string | BasicDataType, source: LineType): BasicDataType | never {
    if (name instanceof BasicDataType) {
      return name;
    }

    const variable = this.memoryAllocation.find((x) => x.name === name);

    if (variable === undefined) {
      this.braingoat.throwError(ErrorType.CompileError, `Variable ${name} is not defined`, source);
    }

    return variable;
  }

  callFunction(name: string, args: TokenType[], block: AST[], source: LineType) {
    if (!(name in Functions)) {
      this.braingoat.throwError(ErrorType.CompileError, `${name} is no valid function`, source);
    }

    Functions[name as keyof typeof Functions](this, args, block, source);
  }

  codeBuilder(code: TemplateStringsArray, ...variables: ReadonlyArray<string | number | Int | Function>) {
    code.forEach((c, i) => {
      this.addCode(c);

      if (variables[i]) {
        if (typeof variables[i] === "string") {
          this.addCode(variables[i] as string);
        } else if (typeof variables[i] === "number") {
          this.addCode(this.movePointerTo(variables[i] as number, false));
        } else if (variables[i] instanceof Int) {
          this.addCode(this.movePointerTo((variables[i] as Int).position[0], false));
        } else if (typeof variables[i] === "function") {
          this.addCode((variables[i] as Function)());
        }
      }
    });
  }

  addCode(...code: string[]) {
    this.code += code
      .join("")
      .split("")
      .filter((x) => VALID_CHARS.includes(x))
      .join("");
  }
}
