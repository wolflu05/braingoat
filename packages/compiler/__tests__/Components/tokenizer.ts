import { Tokenizer } from "../../src/Components/Tokenizer";

describe("tokenize", () => {
  test("tokenize single variable", () => {
    const input = ["int i=0"];
    const tokens = Tokenizer.tokenize(input);
    expect(tokens).toMatchSnapshot();
  });

  test("removes whitespace", () => {
    const input = ["int i       =    ", " 0"];
    const tokens = Tokenizer.tokenize(input);
    expect(tokens).toMatchSnapshot();
  });

  test("work with a single token", () => {
    const input = ["int"];
    const tokens = Tokenizer.tokenize(input);
    expect(tokens).toMatchSnapshot();
  });

  test("function call", () => {
    const input = ["add(a, b, c)"];
    const tokens = Tokenizer.tokenize(input);
    expect(tokens).toMatchSnapshot();
  });

  test("function block call", () => {
    const input = `add(a, b, c) {
                      b()
                   }`.split("\n");
    const tokens = Tokenizer.tokenize(input);
    expect(tokens).toMatchSnapshot();
  });

  test("work with long list of tokens", () => {
    const input = `add(a, b, c) {
                      b()
                      d()
                      while(a,v,b) {
                        while(a,v,c) {
                          while(a,v,c) {
                            while(a,v,c) {
                              hans = a
                              b= 1
                            }
                          }
                        }
                      }
                   }`.split("\n");
    const tokens = Tokenizer.tokenize(input);
    expect(tokens).toMatchSnapshot();
  });
});
