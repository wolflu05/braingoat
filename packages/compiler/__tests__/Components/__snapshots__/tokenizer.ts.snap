// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize function block call 1`] = `
Array [
  Object {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "value": "add",
  },
  Object {
    "source": Object {
      "col": 3,
      "line": 0,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 4,
      "line": 0,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 5,
      "line": 0,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 7,
      "line": 0,
    },
    "value": "b",
  },
  Object {
    "source": Object {
      "col": 8,
      "line": 0,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 10,
      "line": 0,
    },
    "value": "c",
  },
  Object {
    "source": Object {
      "col": 11,
      "line": 0,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 13,
      "line": 0,
    },
    "value": "{",
  },
  Object {
    "source": Object {
      "col": 22,
      "line": 1,
    },
    "value": "b",
  },
  Object {
    "source": Object {
      "col": 23,
      "line": 1,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 24,
      "line": 1,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 19,
      "line": 2,
    },
    "value": "}",
  },
]
`;

exports[`tokenize function call 1`] = `
Array [
  Object {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "value": "add",
  },
  Object {
    "source": Object {
      "col": 3,
      "line": 0,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 4,
      "line": 0,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 5,
      "line": 0,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 7,
      "line": 0,
    },
    "value": "b",
  },
  Object {
    "source": Object {
      "col": 8,
      "line": 0,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 10,
      "line": 0,
    },
    "value": "c",
  },
  Object {
    "source": Object {
      "col": 11,
      "line": 0,
    },
    "value": ")",
  },
]
`;

exports[`tokenize removes whitespace 1`] = `
Array [
  Object {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "value": "int",
  },
  Object {
    "source": Object {
      "col": 4,
      "line": 0,
    },
    "value": "i",
  },
  Object {
    "source": Object {
      "col": 12,
      "line": 0,
    },
    "value": "=",
  },
  Object {
    "source": Object {
      "col": 1,
      "line": 1,
    },
    "value": "0",
  },
]
`;

exports[`tokenize tokenize single variable 1`] = `
Array [
  Object {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "value": "int",
  },
  Object {
    "source": Object {
      "col": 4,
      "line": 0,
    },
    "value": "i",
  },
  Object {
    "source": Object {
      "col": 5,
      "line": 0,
    },
    "value": "=",
  },
  Object {
    "source": Object {
      "col": 6,
      "line": 0,
    },
    "value": "0",
  },
]
`;

exports[`tokenize work with a single token 1`] = `
Array [
  Object {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "value": "int",
  },
]
`;

exports[`tokenize work with long list of tokens 1`] = `
Array [
  Object {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "value": "add",
  },
  Object {
    "source": Object {
      "col": 3,
      "line": 0,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 4,
      "line": 0,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 5,
      "line": 0,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 7,
      "line": 0,
    },
    "value": "b",
  },
  Object {
    "source": Object {
      "col": 8,
      "line": 0,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 10,
      "line": 0,
    },
    "value": "c",
  },
  Object {
    "source": Object {
      "col": 11,
      "line": 0,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 13,
      "line": 0,
    },
    "value": "{",
  },
  Object {
    "source": Object {
      "col": 22,
      "line": 1,
    },
    "value": "b",
  },
  Object {
    "source": Object {
      "col": 23,
      "line": 1,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 24,
      "line": 1,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 22,
      "line": 2,
    },
    "value": "d",
  },
  Object {
    "source": Object {
      "col": 23,
      "line": 2,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 24,
      "line": 2,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 22,
      "line": 3,
    },
    "value": "while",
  },
  Object {
    "source": Object {
      "col": 27,
      "line": 3,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 28,
      "line": 3,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 29,
      "line": 3,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 30,
      "line": 3,
    },
    "value": "v",
  },
  Object {
    "source": Object {
      "col": 31,
      "line": 3,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 32,
      "line": 3,
    },
    "value": "b",
  },
  Object {
    "source": Object {
      "col": 33,
      "line": 3,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 35,
      "line": 3,
    },
    "value": "{",
  },
  Object {
    "source": Object {
      "col": 24,
      "line": 4,
    },
    "value": "while",
  },
  Object {
    "source": Object {
      "col": 29,
      "line": 4,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 30,
      "line": 4,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 31,
      "line": 4,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 32,
      "line": 4,
    },
    "value": "v",
  },
  Object {
    "source": Object {
      "col": 33,
      "line": 4,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 34,
      "line": 4,
    },
    "value": "c",
  },
  Object {
    "source": Object {
      "col": 35,
      "line": 4,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 37,
      "line": 4,
    },
    "value": "{",
  },
  Object {
    "source": Object {
      "col": 26,
      "line": 5,
    },
    "value": "while",
  },
  Object {
    "source": Object {
      "col": 31,
      "line": 5,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 32,
      "line": 5,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 33,
      "line": 5,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 34,
      "line": 5,
    },
    "value": "v",
  },
  Object {
    "source": Object {
      "col": 35,
      "line": 5,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 36,
      "line": 5,
    },
    "value": "c",
  },
  Object {
    "source": Object {
      "col": 37,
      "line": 5,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 39,
      "line": 5,
    },
    "value": "{",
  },
  Object {
    "source": Object {
      "col": 28,
      "line": 6,
    },
    "value": "while",
  },
  Object {
    "source": Object {
      "col": 33,
      "line": 6,
    },
    "value": "(",
  },
  Object {
    "source": Object {
      "col": 34,
      "line": 6,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 35,
      "line": 6,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 36,
      "line": 6,
    },
    "value": "v",
  },
  Object {
    "source": Object {
      "col": 37,
      "line": 6,
    },
    "value": ",",
  },
  Object {
    "source": Object {
      "col": 38,
      "line": 6,
    },
    "value": "c",
  },
  Object {
    "source": Object {
      "col": 39,
      "line": 6,
    },
    "value": ")",
  },
  Object {
    "source": Object {
      "col": 41,
      "line": 6,
    },
    "value": "{",
  },
  Object {
    "source": Object {
      "col": 30,
      "line": 7,
    },
    "value": "hans",
  },
  Object {
    "source": Object {
      "col": 35,
      "line": 7,
    },
    "value": "=",
  },
  Object {
    "source": Object {
      "col": 37,
      "line": 7,
    },
    "value": "a",
  },
  Object {
    "source": Object {
      "col": 30,
      "line": 8,
    },
    "value": "b",
  },
  Object {
    "source": Object {
      "col": 31,
      "line": 8,
    },
    "value": "=",
  },
  Object {
    "source": Object {
      "col": 33,
      "line": 8,
    },
    "value": "1",
  },
  Object {
    "source": Object {
      "col": 28,
      "line": 9,
    },
    "value": "}",
  },
  Object {
    "source": Object {
      "col": 26,
      "line": 10,
    },
    "value": "}",
  },
  Object {
    "source": Object {
      "col": 24,
      "line": 11,
    },
    "value": "}",
  },
  Object {
    "source": Object {
      "col": 22,
      "line": 12,
    },
    "value": "}",
  },
  Object {
    "source": Object {
      "col": 19,
      "line": 13,
    },
    "value": "}",
  },
]
`;
